Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -1119,11 +1202,165 @@ ErrorOr<u32> TagTree::read_value(u32 x, u32 y, Function<ErrorOr<bool>()> const&
+
+        if (r > 0) {
+            PPx--;
+            PPy--;
+            precinct_origin /= 2;
+        }
+
+        // B.7
+        // (B-17)
+        // (The r > 0 check was done right above already.)
+        int xcb_prime = min(coding_parameters.code_block_width_exponent, PPx);
+
+        // (B-18)
+        // (The r > 0 check was done right above already.)
+        int ycb_prime = min(coding_parameters.code_block_height_exponent, PPy);
+
+        for (int precinct_y_index = 0; precinct_y_index < num_precincts_high; ++precinct_y_index) {
+            for (int precinct_x_index = 0; precinct_x_index < num_precincts_wide; ++precinct_x_index) {
+                auto precinct_rect = IntRect({ precinct_x_index * (1 << PPx), precinct_y_index * (1 << PPy), 1 << PPx, 1 << PPy });
+                precinct_rect.set_location(precinct_rect.location() + precinct_origin);
+
+                sub_band.precincts.append(TRY(make_precinct(sub_band, precinct_rect, xcb_prime, ycb_prime)));
+            }
+        }
+
+        return {};
+    };
+
+    auto make_component = [&](TileData const& tile, int component_index) -> ErrorOr<DecodedTileComponent> {
+        DecodedTileComponent component;
+        component.rect = context.siz.reference_grid_coordinates_for_tile_component(tile.rect, component_index);
+
+        dbgln_if(JPEG2000_DEBUG, "making nLL for component {}", component_index);
+        TRY(make_sub_band(tile, component_index, component.nLL, JPEG2000::SubBand::HorizontalLowpassVerticalLowpass, 0));
+
+        auto N_L = context.coding_style_parameters_for_component(tile, component_index).number_of_decomposition_levels;
+        for (int resolution_level = 1; resolution_level <= N_L; ++resolution_level) {
+            DecodedTileComponent::DecodedSubBands sub_bands;
+            for (auto const& [sub_band_index, sub_band] : enumerate(DecodedTileComponent::SubBandOrder)) {
+                dbgln_if(JPEG2000_DEBUG, "r {} making sub-band {} for component {}", resolution_level, (int)sub_band, component_index);
+                TRY(make_sub_band(tile, component_index, sub_bands[sub_band_index], sub_band, resolution_level));
+            }
+            component.decompositions.append(move(sub_bands));
+        }
+
+        return component;
+    };
+
+    auto make_tile = [&](size_t tile_index, TileData& tile) -> ErrorOr<void> {
+        auto const& cod = tile.cod.value_or(context.cod);
+
+        if (cod.may_use_SOP_marker)
+            return Error::from_string_literal("JPEG2000ImageDecoderPlugin: SOP marker not yet implemented");
+        if (cod.may_use_EPH_marker)
+            return Error::from_string_literal("JPEG2000ImageDecoderPlugin: EPH marker not yet implemented");
+
+        auto pq = context.siz.tile_2d_index_from_1d_index(tile_index);
+        tile.rect = context.siz.reference_grid_coordinates_for_tile(pq);
+
+        for (auto const& [component_index, component] : enumerate(context.siz.components)) {
```

Comment:
but in this case I am not sure, but I think `enumerate` Iterator should use a reference internally, so `auto` should suffice (?)

Reply:
According to clangd, just auto is enough and `component` is still a reference. Done, thanks.